Add support for the Web Extensions Event API in the WebProcess.

kiaraarose opened the pull request:
#### f14203440c9ccb1f09af72288eec9fb1b75f235d
<pre>
Add support for the Web Extensions Event API in the WebProcess.
<a href="https://bugs.webkit.org/show_bug.cgi?id=248684">https://bugs.webkit.org/show_bug.cgi?id=248684</a>

Reviewed by Timothy Hatcher.

This patch adds support for the Web Extensions Event API in the WebProcess. This includes support
for adding, removing, and invoking event listeners.

Testing:
Using a test extension, I added a &apos;testEvent&apos; property to the WebExtensionTest API, and I called
&quot;browser.test.testEvent.addListener()&quot; in the background page of the extension. Using the debugger,
I confirmed that the code path for adding the listener hits and a call to
Messages::WebExtensionContext::AddListener is made.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:

* Source/WebKit/Scripts/webkit/messages.py:
(types_that_cannot_be_forward_declared):
Add WebKit::WebExtensionEventListenerType to types that cannot be forward declared so that the
correct type will be used in WebExtensionMessageReceiver.

* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIEventCocoa.mm:
Copied from Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in.
(WebKit::WebExtensionContext::addListener):
(WebKit::WebExtensionContext::removeListener):
Methods for adding and removing event listeners for extensions in the UIProcess.
Implementation being tracked in <a href="https://bugs.webkit.org/show_bug.cgi?id=248684.">https://bugs.webkit.org/show_bug.cgi?id=248684.</a>

* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIEventCocoa.mm: Added.
(WebKit::WebExtensionAPIEvent::invokeListeners):
(WebKit::WebExtensionAPIEvent::invokeListenersWithArgument):
(WebKit::WebExtensionAPIEvent::addListener):
(WebKit::WebExtensionAPIEvent::removeListener):
(WebKit::WebExtensionAPIEvent::hasListener):

* Source/WebKit/WebProcess/Extensions/Bindings/Scripts/CodeGeneratorExtensions.pm:
(_platformTypeVariableDeclaration):
Remove this line of code so that the generated code in JSWebExtensionAPIEvent::addListener can
successfully set the listener. With this line, RefPtr&lt;WebExtensionCallbackHandler&gt; listener
never gets set.

* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIEvent.idl:
Use [NeedsPage] instead of NeedsPageWithCallbackHandler since hasListeners doesn&apos;t need the WebPage.

Canonical link: <a href="https://commits.webkit.org/257375@main">https://commits.webkit.org/257375@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ad401b4be2a7ddb5795fe56264832d8d1d8cd37a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98569 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/7779 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/31693 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/107991 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/8289 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/85157 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91103 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/105975 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/104251 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/6282 "Found 2 new test failures: css3/supports-dom-api.html, fast/text/text-edge-no-half-leading-simple.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/89838 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33277 "Found 1 new test failure: fast/text/text-edge-no-half-leading-simple.html (failure)") | 
| [❌ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/102108 "Found 1 webkitpy python2 test failure: webkitpy.port.server_process_unittest.TestServerProcess.test_broken_pipe") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/88100 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/21197 "Found 4 new test failures: css3/supports-dom-api.html, fast/text/text-edge-no-half-leading-simple.html, fast/text/text-edge-no-half-leading-with-line-height-simple.html, fast/text/text-edge-with-margin-padding-border-simple.html (failure)") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/76210 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/1712 "Built successfully") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/22726 "Found 2 new test failures: css3/supports-dom-api.html, fast/text/text-edge-with-margin-padding-border-simple.html (failure)") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1625 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/45224 "Found 3 new test failures: fast/text/text-edge-with-margin-padding-border-simple.html, imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-dialog-element/modal-dialog-selection.html, media/video-inaccurate-duration-ended.html (failure)") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6563 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42151 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2554 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/3012 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIEvent.h:
```diff
@@ -0,0 +1,66 @@
+public:
+    using ListenerVector = Vector<RefPtr<WebExtensionCallbackHandler>>;
+
+    WebExtensionAPIEvent& event(WebExtensionEventListenerType type) { m_type = type; return *this; }
```
Figured there should be a way for callers to set the event listener type, but I'm not sure if this is the best approach.

xeenon replied:
You are right, we need a way and the constructor is auto generated. Hmm.

How was this being used when you tested it? `create(...).event(Foo)`?

I'll think about this, but for now let's just leave this out.

kiaraarose replied:
Yea I did this:
`m_event = WebExtensionAPIEvent::create(forMainWorld(), runtime(), extensionContext());`
`m_event->event(WebExtensionEventListenerType::ActionOnClicked);`

kiaraarose replied:
But I'll leave this out.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIEventCocoa.mm:
```diff
@@ -0,0 +1,114 @@
+
+    for (auto& listener : m_listeners)
+        listener->call();
+
```
Can remove this line.

kiaraarose committed:
Add support for the Web Extensions Event API in the WebProcess.
https://bugs.webkit.org/show_bug.cgi?id=248684

Reviewed by Timothy Hatcher.

This patch adds support for the Web Extensions Event API in the WebProcess. This includes support
for adding, removing, and invoking event listeners.

Testing:
Using a test extension, I added a 'testEvent' property to the WebExtensionTest API, and I called
"browser.test.testEvent.addListener()" in the background page of the extension. Using the debugger,
I confirmed that the code path for adding the listener hits and a call to
Messages::WebExtensionContext::AddListener is made.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:

* Source/WebKit/Scripts/webkit/messages.py:
(types_that_cannot_be_forward_declared):
Add WebKit::WebExtensionEventListenerType to types that cannot be forward declared so that the
correct type will be used in WebExtensionMessageReceiver.

* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIEventCocoa.mm:
Copied from Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in.
(WebKit::WebExtensionContext::addListener):
(WebKit::WebExtensionContext::removeListener):
Methods for adding and removing event listeners for extensions in the UIProcess.
Implementation being tracked in https://bugs.webkit.org/show_bug.cgi?id=248684.

* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIEventCocoa.mm: Added.
(WebKit::WebExtensionAPIEvent::invokeListeners):
(WebKit::WebExtensionAPIEvent::invokeListenersWithArgument):
(WebKit::WebExtensionAPIEvent::addListener):
(WebKit::WebExtensionAPIEvent::removeListener):
(WebKit::WebExtensionAPIEvent::hasListener):

* Source/WebKit/WebProcess/Extensions/Bindings/Scripts/CodeGeneratorExtensions.pm:
(_platformTypeVariableDeclaration):
Remove this line of code so that the generated code in JSWebExtensionAPIEvent::addListener can
successfully set the listener. With this line, RefPtr<WebExtensionCallbackHandler> listener
never gets set.

* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIEvent.idl:
Use [NeedsPage] instead of NeedsPageWithCallbackHandler since hasListeners doesn't need the WebPage.

Canonical link: https://commits.webkit.org/257375@main
